### PR TITLE
ci(lighthouse): add lighthouse to the ci proof of concept

### PR DESCRIPTION
**Pull request by jzabala**
This PR adds Lighthouse to the CI workflow. Only the collective page is being checked.

Closes https://github.com/opencollective/opencollective/issues/2652

### CODE REVIEW

**Comment by jeffin143** [comment]
Screenshot_2019-11-29-00-07-17-893_com android chrome
 
Didn't the assertion fail , and still circle ci passed ?

**Comment by jzabala** [comment]
> Screenshot_2019-11-29-00-07-17-893_com android chrome
> 
> Didn't the assertion fail , and still circle ci passed ?

For the performance assertion I am using **warm** as the error level. Have a look at the file `lighthouserc.json`. I did that because during test I didn't get a consistent result from it.

**Comment by jeffin143** [comment]
> @jeffin143 For the _performance_ assertion I am using _warm_ as the error level. Have a look at the file `lighthouserc.json`. I did that because during test I didn't get a consistent result from it. This way we avoid false positives.

Fair enough :)

File: package.json [hunk 1]
```diff
@@ -156,7 +156,8 @@
     "test:e2e:2": "CYPRESS_TEST_FILES=2*.js npm run cypress:run",
     "test:e2e:3": "CYPRESS_TEST_FILES=3*.js npm run cypress:run",
     "test:jest": "cross-env TZ=UTC jest components lib pages",
-    "test:update": "npm run test:jest -- --updateSnapshot"
+    "test:update": "npm run test:jest -- --updateSnapshot",
+    "lighthouse": "lhci autorun"
```
Comment (znarf) [inline package.json]: Maybe not useful if it's not available as devDependency.

File: scripts/lighthouse.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+RETURN_CODE=0
+
+if [ "$NODE_ENV" = "circleci" ]; then
+  . ./scripts/setup_functions.sh
+  start_app
+  wait_for_app_services
+
+  echo ""
+  echo "> Running leighthouse"
+  npm run lighthouse
```
Comment (znarf) [inline scripts/lighthouse.sh]: Maybe use `npx` instead.

File: scripts/lighthouse.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+RETURN_CODE=0
+
+if [ "$NODE_ENV" = "circleci" ]; then
+  . ./scripts/setup_functions.sh
+  start_app
+  wait_for_app_services
+
+  echo ""
+  echo "> Running leighthouse"
```
Comment (znarf) [inline scripts/lighthouse.sh]:
```suggestion
  echo "> Running lighthouse"
```

File: .circleci/config.yml [hunk 4]
```diff
@@ -190,6 +195,22 @@ jobs:
...
+      - image: circleci/postgres:9.6.8-alpine-postgis-ram
+      - image: circleci/redis
+      - image: memcached
+    environment:
+      TZ: UTC
+      NODE_ENV: circleci
+    steps:
+      - setup_app
+
+      - run: sudo npm install -g @lhci/cli@0.3.x
```
Comment (znarf) [inline .circleci/config.yml]: Maybe not needed if using npx instead.

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,4 +1,4 @@
-version: 2
+version: 2.1
```
Comment (znarf) [inline .circleci/config.yml]: What do we get with that?

**Comment by znarf** [comment]
A small concern about this PR is that it's introducing a new job, bringing the total from 8 to 9 while having a concurrency of 4. This will make all our Frontend CI 3 minutes longer all the time.

CircleCI is expensive and we don't plan moving to a bigger plan.

Some possibilities:
- do nothing
- redistribute e2e so it's only on 3 different jobs
- run lighthouse together with an e2e test job

File: package.json (hunk 1 above)
Comment (jzabala) [reply package.json]: Done.

File: scripts/lighthouse.sh (hunk 2 above)
Comment (jzabala) [reply scripts/lighthouse.sh]: Done.

File: scripts/lighthouse.sh (hunk 3 above)
Comment (jzabala) [reply scripts/lighthouse.sh]: Great catch.

**Comment by jzabala** [comment]
@znarf great comment about adding lhci as a dev dependency and using npx to run it 👍🏼

> A small concern about this PR is that it's introducing a new job, bringing the total from 8 to 9 while having a concurrency of 4. This will make all our Frontend CI 3 minutes longer all the time.

About this I don't know the context in which was decided to divide the e2e tests in different jobs but why not to run them in one job?

File: .circleci/config.yml (hunk 5 above)
Comment (jzabala) [reply .circleci/config.yml]: [The version 2.1 introduces commands](https://discuss.circleci.com/t/circleci-2-1-config-overview/26057), that I am using to separate the app setup steps from the steps to run e2e tests.

File: .circleci/config.yml (hunk 4 above)
Comment (jzabala) [reply .circleci/config.yml]: Fixed.

File: .circleci/config.yml [hunk 6]
```diff
@@ -17,7 +17,9 @@ aliases:
       paths:
         - ../.npm
 
-  e2e_steps: &e2e_steps
+commands:
```
Comment (jzabala) [inline .circleci/config.yml]: @znarf here is the `commands` stanza that I mention in my comment https://github.com/opencollective/opencollective-frontend/pull/3062#discussion_r352174632

**Comment by znarf** [comment]
> I don't know the context in which was decided to divide the e2e tests in different jobs but why not to run them all in one job?

For parallelization, to get the results faster. Today, we have something like ~5 minutes total parallelized in 4 jobs, while we would have ~12 minutes in a single job. We save around ~7 minutes.

**Comment by jzabala** [comment]
> > I don't know the context in which was decided to divide the e2e tests in different jobs but why not to run them all in one job?
> 
> For parallelization, to get the results faster. Today, we have something like ~5 minutes total parallelized in 4 jobs, while we would have ~12 minutes in a single job. We save around ~7 minutes.

Got it. Those are good numbers. Let me change the script to run Lighthouse with e2e tests.

**Comment by jzabala** [comment]
@znarf done. Lighthouse and e2e tests 0 running together.

**Comment by znarf** [comment]
@jzabala In the end I don't like the solution of mixing lightouse and e2e-0. 

I'm about to introduce GitHub Actions for CI, I think it would be great to have the lighthouse job configured here and not on CircleCI. No concurrency issue that way.

See PR: https://github.com/opencollective/opencollective-frontend/pull/3111 and RFC: https://github.com/opencollective/opencollective/blob/rfc/github-actions/rfcs/008-github-actions.md

**Pull request closed without merging**